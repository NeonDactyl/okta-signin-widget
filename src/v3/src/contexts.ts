/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxMessage, IdxTransaction, OktaAuth } from '@okta/okta-auth-js';
import { createContext } from 'preact';
import { StateUpdater, useContext } from 'preact/hooks';

import {
  FormBag,
  JsonObject,
  Nullish,
  Undefinable,
} from './types';

type IWidgetContext = {
  data: JsonObject;
  authClient: OktaAuth;
  setIdxTransaction: StateUpdater<Nullish<IdxTransaction>>;
  setMessages: StateUpdater<IdxMessage[]>;
  // // TODO: OKTA-502849 - Update param type
  // (RenderSuccessCallback / RenderErrorCallback) once merged into okta-signin-widget
  onSuccessCallback?: (data: JsonObject) => void;
  onErrorCallback?: (data: JsonObject) => void;
  formBag?: FormBag;
  transaction: Nullish<IdxTransaction>;
};

const createWidgetContext = <T extends unknown>() => {
  // Create a context with a generic parameter or undefined
  const genericContext = createContext<Undefinable<T>>(undefined);

  // Check if the value provided to the context is defined or throw an error
  const useWidgetContext = () => {
    const contextIsDefined = useContext(genericContext);
    if (!contextIsDefined) {
      throw new Error('useWidgetContext must be used within a Provider');
    }
    return contextIsDefined;
  };

  return [useWidgetContext, genericContext.Provider] as const;
};

const [useWidgetContext, WidgetContextProvider] = createWidgetContext<IWidgetContext>();

export { useWidgetContext, WidgetContextProvider };
