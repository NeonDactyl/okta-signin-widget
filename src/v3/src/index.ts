/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import './i18n';

import OktaSignIn from './OktaSignIn';

if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
  import('./mocks/browser')
    .then(({ getWorker }) => getWorker())
    .then((worker) => worker?.start());
}

if (typeof window !== 'undefined') {
  // @ts-ignore OKTA-487668
  window.OktaSignIn = OktaSignIn;
}
