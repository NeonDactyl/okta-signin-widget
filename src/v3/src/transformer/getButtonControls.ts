/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ControlElement, JsonSchema7 } from '@jsonforms/core';

import { IDX_STEP } from '../constants';
import { IdxMethod } from '../types';

interface GetButtonControls {
  elements: ControlElement[];
  properties: {
    [property: string]: JsonSchema7;
  };
}

interface GetButtonControlsArgs {
  stepWithSubmit?: boolean;
  stepWithCancel?: boolean;
  stepWithRegister?: boolean;
  stepWithForgotPassword?: boolean;
  stepWithSignInWithFastPass?: boolean;
  idxMethod?: IdxMethod;
  stepWithUnlockAccount?: boolean;
  verifyWithOther?: boolean;
  backToAuthList?: boolean;
}

export enum ButtonOptionType {
  SUBMIT = 'submit',
  CANCEL = 'cancel',
  REGISTER = 'register',
  FORGOT_PASSWORD = 'forgotPassword',
  SIGN_IN_WITH_FASTPASS = 'signInWithFastPass',
  UNLOCK_ACCOUNT = 'unlockAccount',
  CHANGE_AUTHENTICATOR = 'changeAuthenticator',
}

export const getButtonControls = (
  config: GetButtonControlsArgs,
): GetButtonControls => {
  const elements = [];
  const properties: JsonSchema7['properties'] = {};

  if (config.stepWithSubmit) {
    properties[ButtonOptionType.SUBMIT] = {
      type: 'string',
    };

    const submit: ControlElement = {
      type: 'Control',
      label: 'oie.registration.form.update.submit',
      scope: `#/properties/${ButtonOptionType.SUBMIT}`,
      options: {
        format: 'button',
        type: ButtonOptionType.SUBMIT,
        idxMethod: config.idxMethod,
      },
    };

    elements.push(submit);
  }

  if (config.stepWithRegister) {
    properties[ButtonOptionType.REGISTER] = {
      type: 'string',
    };

    const register: ControlElement = {
      type: 'Control',
      label: 'registration.form.submit',
      scope: `#/properties/${ButtonOptionType.REGISTER}`,
      options: {
        format: 'button',
        type: ButtonOptionType.REGISTER,
        variant: 'clear',
      },
    };

    elements.push(register);
  }

  if (config.stepWithForgotPassword) {
    properties[ButtonOptionType.FORGOT_PASSWORD] = {
      type: 'string',
    };

    const forgotPassword: ControlElement = {
      type: 'Control',
      label: 'forgotpassword',
      scope: `#/properties/${ButtonOptionType.FORGOT_PASSWORD}`,
      options: {
        format: 'button',
        type: ButtonOptionType.FORGOT_PASSWORD,
        variant: 'clear',
      },
    };

    elements.push(forgotPassword);
  }

  if (config.stepWithSignInWithFastPass) {
    properties[ButtonOptionType.SIGN_IN_WITH_FASTPASS] = {
      type: 'string',
    };

    const fastPass: ControlElement = {
      type: 'Control',
      label: 'oktaVerify.button',
      scope: `#/properties/${ButtonOptionType.SIGN_IN_WITH_FASTPASS}`,
      options: {
        format: 'button',
        type: ButtonOptionType.SIGN_IN_WITH_FASTPASS,
        // TODO when adding fastpass support.
        // hook up the actual URL from device challenge remediation object in
        // remedation array. The target URL is driven by the API response directly.
        // @see: https://github.com/okta/okta-signin-widget/blob/8a115ae645c6f541b5bb921fd15d0099443de8b6/src/v2/view-builder/views/signin/SignInWithDeviceOption.js#L29-L42
        deviceChallengeUrl: 'http://www.okta.com',
      },
    };

    elements.push(fastPass);
  }

  if (config.stepWithUnlockAccount) {
    properties[ButtonOptionType.UNLOCK_ACCOUNT] = {
      type: 'string',
    };

    const unlock: ControlElement = {
      type: 'Control',
      label: 'unlockaccount',
      scope: `#/properties/${ButtonOptionType.UNLOCK_ACCOUNT}`,
      options: {
        format: 'button',
        type: ButtonOptionType.UNLOCK_ACCOUNT,
        variant: 'clear',
      },
    };

    elements.push(unlock);
  }

  if (config.verifyWithOther) {
    properties[ButtonOptionType.CHANGE_AUTHENTICATOR] = {
      type: 'string',
    };

    elements.push({
      type: 'Control',
      label: 'oie.verification.switch.authenticator',
      scope: `#/properties/${ButtonOptionType.CHANGE_AUTHENTICATOR}`,
      options: {
        format: 'button',
        type: ButtonOptionType.SUBMIT,
        variant: 'clear',
        idxMethod: 'proceed',
        idxStep: IDX_STEP.SELECT_AUTHENTICATOR_AUTHENTICATE,
        skipValidation: true,
      },
    } as ControlElement);
  }

  if (config.backToAuthList) {
    properties[ButtonOptionType.CHANGE_AUTHENTICATOR] = {
      type: 'string',
    };

    elements.push({
      type: 'Control',
      label: 'oie.enroll.switch.authenticator',
      scope: `#/properties/${ButtonOptionType.CHANGE_AUTHENTICATOR}`,
      options: {
        format: 'button',
        type: ButtonOptionType.SUBMIT,
        variant: 'clear',
        idxMethod: 'proceed',
        idxStep: IDX_STEP.SELECT_AUTHENTICATOR_ENROLL,
        skipValidation: true,
      },
    } as ControlElement);
  }

  if (config.stepWithCancel) {
    properties[ButtonOptionType.CANCEL] = {
      type: 'string',
    };

    const cancel: ControlElement = {
      type: 'Control',
      label: 'goback',
      scope: `#/properties/${ButtonOptionType.CANCEL}`,
      options: {
        format: 'button',
        type: ButtonOptionType.CANCEL,
        variant: 'clear',
      },
    };

    elements.push(cancel);
  }

  return { elements, properties };
};
