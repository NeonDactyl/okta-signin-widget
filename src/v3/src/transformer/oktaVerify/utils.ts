/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { UISchemaElement } from '@jsonforms/core';
import { IdxAuthenticator } from '@okta/okta-auth-js';
import set from 'lodash/set';
import {
  DescriptionElement,
  ListElement,
  QRCodeElement,
  StepperNavButtonConfigAttrs,
  StepperNavButtonConfigDirection,
} from 'src/types';

export const appendDescriptionElements = (
  elements: UISchemaElement[],
  authenticator?: IdxAuthenticator,
): void => {
  if (!authenticator?.contextualData) {
    return;
  }

  const { contextualData } = authenticator;

  if (contextualData.qrcode) {
    const { href } = contextualData.qrcode;
    const { displayName } = authenticator;
    elements.push({
      type: 'List',
      options: {
        items: [
          'oie.enroll.okta_verify.qrcode.step1',
          'oie.enroll.okta_verify.qrcode.step2',
          'oie.enroll.okta_verify.qrcode.step3',
        ],
        type: 'ordered',
      },
    } as ListElement);

    elements.push({
      type: 'QRCode',
      options: {
        label: displayName,
        data: href,
      },
    } as QRCodeElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { email } = contextualData;
  if (email) {
    elements.push({
      type: 'Description',
      options: {
        content: 'next.enroll.okta_verify.email.info',
        contentParams: [email],
      },
    } as DescriptionElement);
    return;
  }

  // @ts-ignore OKTA-496373 - missing props from interface
  const { phoneNumber } = contextualData;
  if (phoneNumber) {
    elements.push({
      type: 'Description',
      options: {
        content: 'next.enroll.okta_verify.sms.info',
        contentParams: [phoneNumber],
      },
    } as DescriptionElement);
  }
};

export const createNavButtonConfig = (hasQrCode: boolean): {
  navButtonsConfig: Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
} => {
  const stepperNavBtnConfig = {
    navButtonsConfig: {
      next: {
        variant: 'clear',
        label: hasQrCode
          ? 'enroll.totp.cannotScan'
          : 'next.enroll.okta_verify.switch.channel.link.text',
      },
    } as Record<StepperNavButtonConfigDirection, StepperNavButtonConfigAttrs>,
  };

  if (hasQrCode) {
    set(stepperNavBtnConfig, 'navButtonsConfig.prev', {
      variant: 'clear',
      label: 'next.enroll.okta_verify.switch.channel.link.text',
    } as StepperNavButtonConfigAttrs);
  }

  return stepperNavBtnConfig;
};

export const getTitleKey = (selectedChannel?: string): string => {
  switch (selectedChannel) {
    case 'email':
      return 'oie.enroll.okta_verify.setup.email.title';
    case 'sms':
      return 'oie.enroll.okta_verify.setup.sms.title';
    default:
      return 'oie.enroll.okta_verify.setup.title';
  }
};
