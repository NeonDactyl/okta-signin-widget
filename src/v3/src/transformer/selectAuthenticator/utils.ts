/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';
import {
  IdxOption,
  IdxRemediationValue,
} from '@okta/okta-auth-js/lib/idx/types/idx-js';

import { AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP, AUTHENTICATOR_KEY } from '../../constants';
import {
  AuthenticatorOptionValue,
  Nullish,
  Option,
  Undefinable,
} from '../../types';

const getAuthenticatorOption = (
  options: IdxOption[],
  authenticatorKey: string,
): Undefinable<IdxOption> => options?.find(
  ({ relatesTo }) => relatesTo?.key === authenticatorKey,
);

const getOptionValue = (
  inputs: Input[],
  key: string,
): Nullish<Input> => inputs?.find(
  ({ name }) => name === key,
);

const buildOktaVerifyOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): Option<AuthenticatorOptionValue>[] => {
  const ovRemediation = getAuthenticatorOption(options, AUTHENTICATOR_KEY.OV);
  const id = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'id')?.value;
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType || !methodType.options?.length) {
    return [];
  }

  return methodType.options.map((option) => ({
    key: AUTHENTICATOR_KEY.OV,
    label: option.label,
    description: isEnroll && AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[AUTHENTICATOR_KEY.OV],
    value: {
      label: isEnroll
        ? 'oie.enroll.authenticator.button.text'
        : 'oie.verify.authenticator.button.text',
      methodType: option.value,
      id,
    },
  } as Option<AuthenticatorOptionValue>));
};

const getAuthenticatorDescription = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): Undefinable<string> => {
  if (!authenticatorKey) {
    return undefined;
  }

  if (isEnroll) {
    return AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[authenticatorKey];
  }

  if (authenticatorKey === AUTHENTICATOR_KEY.PHONE) {
    return getAuthenticatorOption(
      options,
      AUTHENTICATOR_KEY.PHONE,
    // @ts-ignore OKTA-499921 (profile missing from relatesTo interface)
    )?.relatesTo?.profile?.phoneNumber;
  }
  return undefined;
};

const getOnPremDescriptionParams = (
  options: IdxOption[],
  authenticatorKey: string,
  isEnroll?: boolean,
): Undefinable<string[]> => {
  if (!isEnroll || authenticatorKey !== AUTHENTICATOR_KEY.ON_PREM) {
    return undefined;
  }

  const vendorName = getAuthenticatorOption(
    options,
    AUTHENTICATOR_KEY.ON_PREM,
  )?.relatesTo?.displayName || 'oie.on_prem.authenticator.default.vendorName';
  return [vendorName];
};

const formatAuthenticatorOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): Option<AuthenticatorOptionValue>[] => (
  options.map((option: IdxOption) => {
    const authenticatorKey = option.relatesTo?.key as string;
    const id = getOptionValue(option?.value as Input[], 'id')?.value;
    const methodType = getOptionValue(option?.value as Input[], 'methodType')?.value;
    const enrollmentId = getOptionValue(option?.value as Input[], 'enrollmentId')?.value;

    return {
      key: authenticatorKey,
      label: option.label,
      description: getAuthenticatorDescription(
        options,
        authenticatorKey,
        isEnroll,
      ),
      descriptionParams: getOnPremDescriptionParams(
        options,
        authenticatorKey,
        isEnroll,
      ),
      value: {
        key: authenticatorKey,
        label: isEnroll
          ? 'oie.enroll.authenticator.button.text'
          : 'oie.verify.authenticator.button.text',
        id,
        methodType,
        enrollmentId,
      } as AuthenticatorOptionValue,
    };
  })
);

const getAuthenticatorOptions = (
  options: IdxOption[],
  isEnroll?: boolean,
): Option<AuthenticatorOptionValue>[] => {
  const formattedOptions = formatAuthenticatorOptions(options, isEnroll);

  // appending OV options back to its original spot
  const ovOptions = buildOktaVerifyOptions(options, isEnroll);
  if (ovOptions.length && options?.length) {
    const ovIndex = options.findIndex(({ relatesTo }) => relatesTo?.key === AUTHENTICATOR_KEY.OV);
    formattedOptions.splice(ovIndex, 1, ...ovOptions);
  }

  return formattedOptions;
};

export const getOVMethodTypeAuthenticatorOptions = (
  options?: IdxOption[],
): Option<AuthenticatorOptionValue>[] => {
  if (!options) {
    return [];
  }

  return options.map((option) => ({
    key: AUTHENTICATOR_KEY.OV,
    label: option.label,
    value: {
      label: 'oie.verify.authenticator.button.text',
      methodType: option.value as string,
    },
  }));
};

export const isOnlyPushWithAutoChallenge = (
  authenticatorRemediation?: IdxRemediationValue,
): boolean => {
  const methodType = (authenticatorRemediation?.form?.value)
    ?.find(({ name }) => name === 'methodType');
  const autoChallenge = (authenticatorRemediation?.form?.value)
    ?.find(({ name }) => name === 'autoChallenge');

  return typeof autoChallenge !== 'undefined'
    && methodType?.options?.length === 1
    && methodType.options[0].value === 'push';
};

export const getAuthenticatorVerifyOptions = (
  authenticatorOptions: IdxOption[],
): Option<AuthenticatorOptionValue>[] => getAuthenticatorOptions(
  authenticatorOptions,
);

export const getAuthenticatorEnrollOptions = (
  authenticatorOptions: IdxOption[],
): Option<AuthenticatorOptionValue>[] => getAuthenticatorOptions(
  authenticatorOptions,
  true,
);
