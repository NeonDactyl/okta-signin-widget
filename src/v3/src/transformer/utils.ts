/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ControlElement } from '@jsonforms/core';
import { IdxTransaction } from '@okta/okta-auth-js';
import { IdxRemediation } from '@okta/okta-auth-js/lib/idx/types/idx-js';

import { FormBag, LayoutType, RequiredKeys } from '../types';

export const createForm = (): FormBag => ({
  schema: {
    type: 'object',
    properties: {},
    required: [],
  },
  uischema: {
    type: LayoutType.VERTICAL,
    elements: [],
  },
  data: {},
});

export const removeUIElementWithScope = (
  scope: string,
  elements: ControlElement[],
): ControlElement[] => (
  elements.filter((element) => (
    scope !== element.scope
  ))
);

export const getUIElementWithScope = (
  scope: string,
  elements: ControlElement[],
): ControlElement | undefined => (
  elements.find((element) => (
    scope === element.scope
  ))
);

export const isNextStepAvailable = (
  transaction: IdxTransaction,
): transaction is RequiredKeys<IdxTransaction, 'nextStep'> => !!transaction.nextStep;

export const remediationContainsStep = (
  neededToProceed: IdxRemediation[],
  step: string,
): boolean => neededToProceed.some((remediation) => remediation.name === step);

export const getCurrentTimestamp = (): number => (new Date().getTime());
