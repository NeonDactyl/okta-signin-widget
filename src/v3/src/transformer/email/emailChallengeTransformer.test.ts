/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { ControlElement, Layout } from '@jsonforms/core';
import { getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import { FormBag, LayoutType, StepperLayout } from 'src/types';

import { ButtonOptionType } from '../getButtonControls';
import * as utils from '../utils';
import { transformEmailChallenge } from '.';

describe('EmailChallengeTransformer Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  let formBag: FormBag;

  beforeEach(() => {
    jest.spyOn(utils, 'getCurrentTimestamp').mockReturnValue(123456789);
    formBag = {
      envelope: {},
      data: {},
      schema: {},
      uischema: {
        type: 'VerticalLayout',
        elements: [],
      },
    };
  });

  it('should create email challenge UI elements when resend code is available', () => {
    transaction.nextStep = {
      canResend: true,
      authenticator: {
        // @ts-ignore OKTA-483184 (profile missing from authenticator interface)
        profile: {
          email: 'firstname.lastname@example.com',
        },
      },
    };
    const updatedFormBag = transformEmailChallenge(transaction, formBag);

    expect(updatedFormBag).toMatchSnapshot();

    expect(updatedFormBag.uischema.elements.length).toBe(1);
    expect(updatedFormBag.uischema.elements[0].type).toBe(LayoutType.STEPPER);
    expect(updatedFormBag.uischema.elements[0].options?.navButtonsConfig?.next?.variant).toBe('clear');
    expect(updatedFormBag.uischema.elements[0].options?.key).toBeDefined();

    const stepperElements = (updatedFormBag.uischema.elements[0] as StepperLayout).elements;

    const layoutOne = stepperElements[0] as Layout;

    expect(layoutOne.type).toBe(LayoutType.VERTICAL);
    expect(layoutOne.elements.length).toBe(3);

    expect(layoutOne.elements[2].type).toBe('Description');
    expect(layoutOne.elements[2].options?.content).toBe('next.email.challenge.informationalTextWithEmail');
    expect(layoutOne.elements[2].options?.contentParams?.[0]).toBe('firstname.lastname@example.com');

    const layoutTwo = stepperElements[1] as Layout;

    expect(layoutTwo.type).toBe(LayoutType.VERTICAL);
    expect(layoutTwo.elements.length).toBe(5);
    expect(layoutTwo.elements[2].type).toBe('Description');
    expect(layoutTwo.elements[2].options?.content).toBe('next.email.challenge.informationalTextWithEmail');
    expect(layoutTwo.elements[2].options?.contentParams?.[0]).toBe('firstname.lastname@example.com');

    expect(layoutTwo.elements[3].type).toBe('Control');
    expect((layoutTwo.elements[3] as ControlElement).label).toBe('email.enroll.enterCode');

    expect(layoutTwo.elements[4].options?.format).toBe('button');
    expect(layoutTwo.elements[4].options?.type).toBe(ButtonOptionType.SUBMIT);
  });

  it('should create email challenge UI elements when profile email is NOT available', () => {
    transaction.nextStep = {
      canResend: true,
      authenticator: {
        // @ts-ignore OKTA-483184 (profile missing from authenticator interface)
        profile: undefined,
      },
    };
    const updatedFormBag = transformEmailChallenge(transaction, formBag);

    expect(updatedFormBag).toMatchSnapshot();

    const stepperElements = (updatedFormBag.uischema.elements[0] as StepperLayout).elements;

    const layoutOne = stepperElements[0] as Layout;

    expect(layoutOne.elements.length).toBe(3);

    expect(layoutOne.elements[2].options?.content).toBe('next.email.challenge.informationalText');
    expect(layoutOne.elements[2].options?.contentParams?.[0]).toBeUndefined();

    const layoutTwo = stepperElements[1] as Layout;

    expect(layoutTwo.elements.length).toBe(5);
    expect(layoutTwo.elements[2].options?.content).toBe('next.email.challenge.informationalText');
    expect(layoutTwo.elements[2].options?.contentParams?.[0]).toBeUndefined();
  });

  it('should create email challenge UI elements when resend code is NOT available', () => {
    transaction.nextStep = {
      canResend: false,
      authenticator: {
        // @ts-ignore OKTA-483184 (profile missing from authenticator interface)
        profile: {
          email: 'firstname.lastname@example.com',
        },
      },
    };
    const updatedFormBag = transformEmailChallenge(transaction, formBag);

    expect(updatedFormBag).toMatchSnapshot();

    expect(updatedFormBag.uischema.elements.length).toBe(1);
    expect(updatedFormBag.uischema.elements[0].type).toBe(LayoutType.STEPPER);
    expect(updatedFormBag.uischema.elements[0].options?.navButtonsConfig?.next?.variant).toBe('clear');
    expect(updatedFormBag.uischema.elements[0].options?.key).toBeDefined();

    const stepperElements = (updatedFormBag.uischema.elements[0] as StepperLayout).elements;

    const layoutOne = stepperElements[0] as Layout;

    expect(layoutOne.type).toBe(LayoutType.VERTICAL);
    expect(layoutOne.elements.length).toBe(2);

    expect(layoutOne.elements[1].type).toBe('Description');
    expect(layoutOne.elements[1].options?.content).toBe('next.email.challenge.informationalTextWithEmail');
    expect(layoutOne.elements[1].options?.contentParams?.[0]).toBe('firstname.lastname@example.com');

    const layoutTwo = stepperElements[1] as Layout;

    expect(layoutTwo.type).toBe(LayoutType.VERTICAL);
    expect(layoutTwo.elements.length).toBe(4);
    expect(layoutTwo.elements[1].type).toBe('Description');
    expect(layoutTwo.elements[1].options?.content).toBe('next.email.challenge.informationalTextWithEmail');
    expect(layoutTwo.elements[1].options?.contentParams?.[0]).toBe('firstname.lastname@example.com');

    expect(layoutTwo.elements[2].type).toBe('Control');
    expect((layoutTwo.elements[2] as ControlElement).label).toBe('email.enroll.enterCode');

    expect(layoutTwo.elements[3].options?.format).toBe('button');
    expect(layoutTwo.elements[3].options?.type).toBe(ButtonOptionType.SUBMIT);
  });
});
