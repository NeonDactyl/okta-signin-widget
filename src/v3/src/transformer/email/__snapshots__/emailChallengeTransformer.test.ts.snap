// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailChallengeTransformer Tests should create email challenge UI elements when profile email is NOT available 1`] = `
Object {
  "data": Object {},
  "envelope": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "ctaText": "email.code.not.received",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.email.mfa.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "next.email.challenge.informationalText",
                  "contentParams": Array [
                    undefined,
                  ],
                },
                "type": "Description",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "ctaText": "email.code.not.received",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.email.mfa.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "next.email.challenge.informationalText",
                  "contentParams": Array [
                    undefined,
                  ],
                },
                "type": "Description",
              },
              Object {
                "label": "email.enroll.enterCode",
                "scope": "#/properties/verificationCode",
                "type": "Control",
              },
              Object {
                "label": "mfa.challenge.verify",
                "options": Object {
                  "format": "button",
                  "type": "submit",
                },
                "scope": "#/properties/submit",
                "type": "Control",
              },
            ],
            "type": "VerticalLayout",
          },
        ],
        "options": Object {
          "key": "123456789-emailChallenge",
          "navButtonsConfig": Object {
            "next": Object {
              "label": "oie.email.verify.alternate.showCodeTextField",
              "variant": "clear",
            },
          },
        },
        "type": "Stepper",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`EmailChallengeTransformer Tests should create email challenge UI elements when resend code is NOT available 1`] = `
Object {
  "data": Object {},
  "envelope": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "content": "oie.email.mfa.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "next.email.challenge.informationalTextWithEmail",
                  "contentParams": Array [
                    "firstname.lastname@example.com",
                  ],
                },
                "type": "Description",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "content": "oie.email.mfa.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "next.email.challenge.informationalTextWithEmail",
                  "contentParams": Array [
                    "firstname.lastname@example.com",
                  ],
                },
                "type": "Description",
              },
              Object {
                "label": "email.enroll.enterCode",
                "scope": "#/properties/verificationCode",
                "type": "Control",
              },
              Object {
                "label": "mfa.challenge.verify",
                "options": Object {
                  "format": "button",
                  "type": "submit",
                },
                "scope": "#/properties/submit",
                "type": "Control",
              },
            ],
            "type": "VerticalLayout",
          },
        ],
        "options": Object {
          "key": "123456789-emailChallenge",
          "navButtonsConfig": Object {
            "next": Object {
              "label": "oie.email.verify.alternate.showCodeTextField",
              "variant": "clear",
            },
          },
        },
        "type": "Stepper",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`EmailChallengeTransformer Tests should create email challenge UI elements when resend code is available 1`] = `
Object {
  "data": Object {},
  "envelope": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "ctaText": "email.code.not.received",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.email.mfa.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "next.email.challenge.informationalTextWithEmail",
                  "contentParams": Array [
                    "firstname.lastname@example.com",
                  ],
                },
                "type": "Description",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "ctaText": "email.code.not.received",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.email.mfa.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "next.email.challenge.informationalTextWithEmail",
                  "contentParams": Array [
                    "firstname.lastname@example.com",
                  ],
                },
                "type": "Description",
              },
              Object {
                "label": "email.enroll.enterCode",
                "scope": "#/properties/verificationCode",
                "type": "Control",
              },
              Object {
                "label": "mfa.challenge.verify",
                "options": Object {
                  "format": "button",
                  "type": "submit",
                },
                "scope": "#/properties/submit",
                "type": "Control",
              },
            ],
            "type": "VerticalLayout",
          },
        ],
        "options": Object {
          "key": "123456789-emailChallenge",
          "navButtonsConfig": Object {
            "next": Object {
              "label": "oie.email.verify.alternate.showCodeTextField",
              "variant": "clear",
            },
          },
        },
        "type": "Stepper",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
